### PR TITLE
dialects: Add a stencil.CastOp -> memref.Cast lowering.

### PR DESCRIPTION
**Pull request by PapyChacal**
There will be PyRight issues, but I'm opening to get advice: I'm confused by this new typechecking error on ArrayAttr.__new__, and can't remember how to make pyright happy with the type parameter of a MemRefType :thinking:

### CODE REVIEW

**Comment by PapyChacal** [comment]
NVM, I think I got them

**Review by webmiche — APPROVED** [review]
LGTM

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py [hunk 1]
```diff
@@ -1,67 +1,74 @@
...
 def GetMemRefFromFieldWithLBAndUB(memref_element_type: _TypeElement,
                                   lb: IndexAttr,
                                   ub: IndexAttr) -> MemRefType[_TypeElement]:
-    memref_shape_integer_attr_list: list[AnyIntegerAttr] = []
-    for i in range(len(lb.array.data)):
-        memref_shape_integer_attr_list.append(
-            IntegerAttr.from_params(
-                ub.array.data[i].value.data - lb.array.data[i].value.data,
-                IndexType()))
+    dims = [
```
Comment (georgebisbas) [inline xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:29]: I would add a comment on this, although I it seems clear to me

**Comment by PapyChacal** [comment]
Applied the comments, I'll merge as-is as it's all in experimental to iterate!

**Pull request merged**